File: evals/models/analyze.ts
import { readJsonLines } from '../lib/jsonl.ts'
import type { Labelled } from '../run-retrieval.ts'
import { runKey, type Run } from './compare.ts'
import type { Preference } from './judge.ts'

/**
 * Computes the ThinkingCap comparison strictly to evals/results/thinkingcap-preregistration.md and
 * its five amendments. Written after the data was collected and judged, which is only safe because
 * the design was fixed before: every metric and cut below is named in that document, and nothing
 * here was chosen by looking at the numbers first.
 */

/** The two sets are reported separately and never averaged; see the note in compare.ts. */
const SET = process.env.EVAL_SET === 'bulk' ? 'bulk' : 'hard'
/** Must name the same serving conditions the runs were collected under; see compare.ts. */
const RUN_TAG = process.env.EVAL_RUN_TAG ?? 'hfe'
const RUNS = `evals/data/model-runs.${RUN_TAG}.${SET}.jsonl`
/** Which judging pass to score against; the blind one is the primary. See amendment 6. */
const PASS = process.env.EVAL_JUDGE_PASS ?? 'blind'
const PREFS = `evals/data/model-preferences.${RUN_TAG}.${SET}.${PASS}.jsonl`

/** Seed 1 is the pass that was judged, so answers come from it; see amendment 6a. */
export function firstPerPair(runs: Run[]): Run[] {
  const seen = new Set<string>()
  return runs.filter((r) => {
    const k = runKey(r.model, r.query)
    if (seen.has(k)) {
      return false
    }
    seen.add(k)
    return true
  })
}

function bySeedGroup(runs: Run[]): Map<string, Run[]> {
  const byPair = new Map<string, Run[]>()
  for (const r of runs) {
    const k = runKey(r.model, r.query)
    if (!byPair.has(k)) {
      byPair.set(k, [])
    }
    byPair.get(k)!.push(r)
  }
  return byPair
}

/**
 * How far the same model's trace length moves on the same question across seeds. Sampling is what
 * produces this now rather than a resume bug, but it plays the same role: a saving smaller than
 * this floor is a draw dressed as a result.
 */
export function repeatSpreads(runs: Run[]): number[] {
  const out: number[] = []
  for (const rs of bySeedGroup(runs).values()) {
    if (rs.length < 2) {
      continue
    }
    const lens = rs.map((r) => r.reasoningChars)
    const max = Math.max(...lens)
    if (max > 0) {
      out.push((max - Math.min(...lens)) / max)
    }
  }
  return out
}

/**
 * One row per model and question, holding the median across that pair's seeds. The median rather
 * than the mean because a single run that reaches the ceiling would otherwise carry the question
 * with it — see amendment 6a, fixed before the data was read.
 */
export function medianAcrossSeeds(runs: Run[]): Run[] {
  const out: Run[] = []
  for (const rs of bySeedGroup(runs).values()) {
    out.push({
      ...rs[0],
      reasoningChars: median(rs.map((r) => r.reasoningChars)),
      completionTokens: median(rs.map((r) => r.completionTokens ?? Number.NaN)),
    })
  }
  return out
}

export function median(xs: number[]): number {
  if (xs.length === 0) {
    return Number.NaN
  }
  const s = [...xs].sort((a, b) => a - b)
  const mid = s.length >> 1
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2
}

function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a = (a + 0x6d2b79f5) >>> 0
    let t = Math.imul(a ^ (a >>> 15), 1 | a)
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/**
 * Paired bootstrap over the per-question values: questions are resampled, not the two models
 * independently, because the pairing is what removes question difficulty from the comparison.
 */
export function bootstrapMedian(values: number[], seed = 7, rounds = 10000): [number, number] {
  if (values.length === 0) {
    return [Number.NaN, Number.NaN]
  }
  const rand = mulberry32(seed)
  const medians: number[] = []
  for (let i = 0; i < rounds; i++) {
    const sample: number[] = []
    for (let j = 0; j < values.length; j++) {
      sample.push(values[Math.floor(rand() * values.length)])
    }
    medians.push(median(sample))
  }
  medians.sort((a, b) => a - b)
  return [medians[Math.floor(rounds * 0.025)], medians[Math.floor(rounds * 0.975)]]
}

/**
 * Wilson interval rather than the normal approximation: with ~20 non-tied pairs the normal one
 * runs off the end of [0,1] and would report an interval that cannot contain the true value.
 */
export function wilson(successes: number, n: number, z = 1.96): [number, number] {
  if (n === 0) {
    return [Number.NaN, Number.NaN]
  }
  const p = successes / n
  const d = 1 + (z * z) / n
  const centre = p + (z * z) / (2 * n)
  const spread = z * Math.sqrt((p * (1 - p)) / n + (z * z) / (4 * n * n))
  return [(centre - spread) / d, (centre + spread) / d]
}

type Paired = {
  query: string
  base: Run
  tuned: Run
  /** Whether the context the two models shared held any chunk the labeller called relevant. */
  hasRelevantContext: boolean
}

function pct(x: number): string {
  return `${(x * 100).toFixed(1)}%`
}

function main(): void {
  const baseModel = process.env.BASE_MODEL
  const tunedModel = process.env.TUNED_MODEL
  if (!baseModel || !tunedModel) {
    throw new Error('set BASE_MODEL and TUNED_MODEL')
  }

  const allRuns = readJsonLines<Run>(RUNS, { required: true })
  // Lengths are the seed medians; every other field on the row is seed 1's, which is the pass that
  // was judged, so the answers quoted below are the answers that were actually compared.
  const runs = medianAcrossSeeds(allRuns)
  const prefs = readJsonLines<Preference>(PREFS, { required: true })
  const labels = readJsonLines<Labelled>('evals/data/retrieval.labels.jsonl', { required: true })
  const relevantByQuery = new Map(labels.map((l) => [l.query, new Set(l.relevant)]))

  const pairs: Paired[] = []
  for (const query of new Set(runs.map((r) => r.query))) {
    const base = runs.find((r) => r.query === query && r.model === baseModel)
    const tuned = runs.find((r) => r.query === query && r.model === tunedModel)
    if (!base || !tuned) {
      continue
    }
    const rel = relevantByQuery.get(query) ?? new Set<string>()
    pairs.push({ query, base, tuned, hasRelevantContext: base.contextIds.some((id) => rel.has(id)) })
  }

  console.log(`${SET} set — n = ${pairs.length} otázek, oba modely, byte-identický vstup\n`)

  // Primary metric (amendment 3): trace length in characters. The token count is not comparable
  // across these two providers and is excluded, not quietly used.
  const charRed = pairs.map((p) => (p.base.reasoningChars - p.tuned.reasoningChars) / p.base.reasoningChars)
  const [cLo, cHi] = bootstrapMedian(charRed)
  console.log('## Primární metrika — délka uvažovací stopy (znaky)')
  console.log(`base medián    ${median(pairs.map((p) => p.base.reasoningChars)).toFixed(0)} zn`)
  console.log(`tuned medián   ${median(pairs.map((p) => p.tuned.reasoningChars)).toFixed(0)} zn`)
  console.log(`úspora mediánu ${pct(median(charRed))}  95% CI [${pct(cLo)}, ${pct(cHi)}]`)
  console.log(`kratší stopa u ${charRed.filter((x) => x > 0).length}/${pairs.length} otázek`)

  const spreads = repeatSpreads(allRuns)
  console.log(`\npodlaha šumu (rozptyl mezi seedy): medián ${pct(median(spreads))}, n = ${spreads.length}`)
  if (median(charRed) <= median(spreads)) {
    console.log('POZOR: úspora nepřesahuje podlahu šumu — viz dodatek 6a.')
  }

  // Secondary: completion tokens are trustworthy on both sides and are the practical cost measure.
  const tokRed = pairs
    .filter((p) => p.base.completionTokens && p.tuned.completionTokens)
    .map((p) => (p.base.completionTokens! - p.tuned.completionTokens!) / p.base.completionTokens!)
  const [tLo, tHi] = bootstrapMedian(tokRed, 11)
  console.log('\n## Sekundární — completion tokeny')
  console.log(`base medián    ${median(pairs.map((p) => p.base.completionTokens ?? Number.NaN)).toFixed(0)}`)
  console.log(`tuned medián   ${median(pairs.map((p) => p.tuned.completionTokens ?? Number.NaN)).toFixed(0)}`)
  console.log(`úspora mediánu ${pct(median(tokRed))}  95% CI [${pct(tLo)}, ${pct(tHi)}]`)

  const untrustworthy = pairs
    .flatMap((p) => [p.base, p.tuned])
    .filter((r) => r.reasoningTokens === null || (r.reasoningTokens === 0 && r.reasoningChars > 0)).length
  console.log(`\nreasoning_tokens nedůvěryhodné u ${untrustworthy}/${pairs.length * 2} běhů — vyloučeno (dodatek 3)`)

  // Quality: the saving means nothing unless capability held.
  console.log('\n## Kvalita — slepé párové posouzení')
  const verdicts = pairs.map((p) => ({ p, w: prefs.find((x) => x.query === p.query)?.winner }))
  const judged = verdicts.filter((v) => v.w !== undefined)
  const ties = judged.filter((v) => v.w === 'tie').length
  const baseWins = judged.filter((v) => v.w === baseModel).length
  const tunedWins = judged.filter((v) => v.w === tunedModel).length
  const decided = baseWins + tunedWins
  const [pLo, pHi] = wilson(baseWins, decided)
  console.log(`posouzeno ${judged.length}, remíz ${ties}, rozhodnutých ${decided}`)
  console.log(`base ${baseWins} : ${tunedWins} tuned`)
  console.log(`podíl pro base ${decided ? pct(baseWins / decided) : '—'}  95% CI [${pct(pLo)}, ${pct(pHi)}]`)
  const detectable = decided >= 20
  if (pLo <= 0.5 && pHi >= 0.5) {
    console.log(
      `→ interval obsahuje 0.5 — rozdíl neprokázán${detectable ? '' : `, ale ${decided} rozhodnutých dvojic ho neumí odhalit`}`,
    )
  } else {
    console.log('→ interval míjí 0.5 — rozdíl v kvalitě')
  }

  // Not in the original design, added because collection revealed it: on 12 of 30 questions the
  // shared context held no chunk the labeller called relevant, so those test refusal, not recall.
  console.log('\n## Rozpad podle toho, jestli kontext vůbec obsahoval odpověď')
  for (const withCtx of [true, false]) {
    const group = judged.filter((v) => v.p.hasRelevantContext === withCtx)
    const t = group.filter((v) => v.w === 'tie').length
    const b = group.filter((v) => v.w === baseModel).length
    const u = group.filter((v) => v.w === tunedModel).length
    console.log(
      `${withCtx ? 's relevantním kontextem ' : 'bez relevantního      '} n=${group.length}  base ${b} : ${u} tuned, remíz ${t}`,
    )
  }

  // The promised blindness check: if verdicts tracked answer length, the judging measured length.
  console.log('\n## Kontrola sleposti — šel verdikt za délkou odpovědi?')
  const decidedPairs = judged.filter((v) => v.w !== 'tie')
  const longerWon = decidedPairs.filter((v) => {
    const winner = v.w === baseModel ? v.p.base : v.p.tuned
    const loser = v.w === baseModel ? v.p.tuned : v.p.base
    return winner.answer.length > loser.answer.length
  }).length
  const [lLo, lHi] = wilson(longerWon, decidedPairs.length)
  console.log(`delší odpověď vyhrála ${longerWon}/${decidedPairs.length}  95% CI [${pct(lLo)}, ${pct(lHi)}]`)
  console.log(
    `→ ${lLo <= 0.5 && lHi >= 0.5 ? 'nerozlišitelné od náhody — délka verdikt neřídila' : 'délka s verdiktem koreluje — nahlásit jako limit'}`,
  )
}

if (/[\\/]models[\\/]analyze\.ts$/.test(process.argv[1] ?? '')) {
  main()
}

File: evals/models/analyze.py
import math
import os
from typing import NamedTuple

from lib.env import require_env
from lib.jsonl import read_json_lines
from lib.metrics import mulberry32
from lib.paths import DATA
from models.compare import Run, run_key
from models.judge import Preference
from run_retrieval import Labelled

SET = 'bulk' if os.environ.get('EVAL_SET') == 'bulk' else 'hard'
RUN_TAG = os.environ.get('EVAL_RUN_TAG', 'hfe')
RUNS = DATA / f'model-runs.{RUN_TAG}.{SET}.jsonl'
PASS = os.environ.get('EVAL_JUDGE_PASS', 'blind')
PREFS = DATA / f'model-preferences.{RUN_TAG}.{SET}.{PASS}.jsonl'


def first_per_pair(runs: list[Run]) -> list[Run]:
    seen: set[str] = set()
    out: list[Run] = []
    for r in runs:
        k = run_key(r['model'], r['query'])
        if k not in seen:
            seen.add(k)
            out.append(r)
    return out


def by_seed_group(runs: list[Run]) -> dict[str, list[Run]]:
    by_pair: dict[str, list[Run]] = {}
    for r in runs:
        by_pair.setdefault(run_key(r['model'], r['query']), []).append(r)
    return by_pair


def repeat_spreads(runs: list[Run]) -> list[float]:
    out: list[float] = []
    for rs in by_seed_group(runs).values():
        if len(rs) < 2:
            continue
        lens = [r['reasoningChars'] for r in rs]
        if max(lens) > 0:
            out.append((max(lens) - min(lens)) / max(lens))
    return out


def median(xs: list[float]) -> float:
    if not xs:
        return math.nan
    s = sorted(xs)
    mid = len(s) // 2
    return s[mid] if len(s) % 2 else (s[mid - 1] + s[mid]) / 2


def median_across_seeds(runs: list[Run]) -> list[Run]:
    out: list[Run] = []
    for rs in by_seed_group(runs).values():
        tokens = [r['completionTokens'] if r['completionTokens'] is not None else math.nan for r in rs]
        merged: Run = {
            **rs[0],
            'reasoningChars': int(median([r['reasoningChars'] for r in rs])),
            'completionTokens': int(median(tokens)) if not math.isnan(median(tokens)) else None,
        }
        out.append(merged)
    return out


def bootstrap_median(values: list[float], seed: int = 7, rounds: int = 10_000) -> tuple[float, float]:
    if not values:
        return math.nan, math.nan
    rand = mulberry32(seed)
    n = len(values)
    medians = sorted(median([values[math.floor(rand() * n)] for _ in range(n)]) for _ in range(rounds))
    return medians[math.floor(rounds * 0.025)], medians[math.floor(rounds * 0.975)]


def wilson(successes: int, n: int, z: float = 1.96) -> tuple[float, float]:
    if n == 0:
        return math.nan, math.nan
    p = successes / n
    d = 1 + z * z / n
    centre = p + z * z / (2 * n)
    spread = z * math.sqrt(p * (1 - p) / n + z * z / (4 * n * n))
    return (centre - spread) / d, (centre + spread) / d


class Paired(NamedTuple):
    query: str
    base: Run
    tuned: Run
    has_relevant_context: bool


def pct(x: float) -> str:
    return f'{x * 100:.1f}%'


def interval(lo: float, hi: float) -> str:
    return f'[{pct(lo)}, {pct(hi)}]'


def pair_up(runs: list[Run], base_model: str, tuned_model: str, labels: list[Labelled]) -> list[Paired]:
    relevant_by_query = {label['query']: set(label['relevant']) for label in labels}
    pairs: list[Paired] = []
    for query in dict.fromkeys(r['query'] for r in runs):
        base = next((r for r in runs if r['query'] == query and r['model'] == base_model), None)
        tuned = next((r for r in runs if r['query'] == query and r['model'] == tuned_model), None)
        if base is None or tuned is None:
            continue
        rel = relevant_by_query.get(query, set())
        pairs.append(Paired(query, base, tuned, any(id in rel for id in base['contextIds'])))
    return pairs


def report_lengths(pairs: list[Paired], all_runs: list[Run]) -> None:
    char_red = [
        (p.base['reasoningChars'] - p.tuned['reasoningChars']) / p.base['reasoningChars'] for p in pairs
    ]
    c_lo, c_hi = bootstrap_median(char_red)
    print('## Primární metrika — délka uvažovací stopy (znaky)')
    print(f'base medián    {median([p.base["reasoningChars"] for p in pairs]):.0f} zn')
    print(f'tuned medián   {median([p.tuned["reasoningChars"] for p in pairs]):.0f} zn')
    print(f'úspora mediánu {pct(median(char_red))}  95% CI {interval(c_lo, c_hi)}')
    print(f'kratší stopa u {sum(1 for x in char_red if x > 0)}/{len(pairs)} otázek')

    spreads = repeat_spreads(all_runs)
    print(f'\npodlaha šumu (rozptyl mezi seedy): medián {pct(median(spreads))}, n = {len(spreads)}')
    if median(char_red) <= median(spreads):
        print('POZOR: úspora nepřesahuje podlahu šumu — viz dodatek 6a.')

    with_tokens = [p for p in pairs if p.base['completionTokens'] and p.tuned['completionTokens']]
    tok_red = [
        (p.base['completionTokens'] - p.tuned['completionTokens']) / p.base['completionTokens']
        for p in with_tokens
        if p.base['completionTokens'] is not None and p.tuned['completionTokens'] is not None
    ]
    t_lo, t_hi = bootstrap_median(tok_red, 11)
    print('\n## Sekundární — completion tokeny')
    print(f'base medián    {median([tokens_or_nan(p.base) for p in pairs]):.0f}')
    print(f'tuned medián   {median([tokens_or_nan(p.tuned) for p in pairs]):.0f}')
    print(f'úspora mediánu {pct(median(tok_red))}  95% CI {interval(t_lo, t_hi)}')

    untrustworthy = sum(
        1
        for r in (run for p in pairs for run in (p.base, p.tuned))
        if r['reasoningTokens'] is None or (r['reasoningTokens'] == 0 and r['reasoningChars'] > 0)
    )
    print(f'\nreasoning_tokens nedůvěryhodné u {untrustworthy}/{len(pairs) * 2} běhů — vyloučeno (dodatek 3)')


def report_quality(pairs: list[Paired], prefs: list[Preference], base_model: str, tuned_model: str) -> None:
    print('\n## Kvalita — slepé párové posouzení')
    winner_of = {p['query']: p['winner'] for p in prefs}
    judged = [(p, winner_of[p.query]) for p in pairs if p.query in winner_of]
    ties = sum(1 for _, w in judged if w == 'tie')
    base_wins = sum(1 for _, w in judged if w == base_model)
    tuned_wins = sum(1 for _, w in judged if w == tuned_model)
    decided = base_wins + tuned_wins
    p_lo, p_hi = wilson(base_wins, decided)
    print(f'posouzeno {len(judged)}, remíz {ties}, rozhodnutých {decided}')
    print(f'base {base_wins} : {tuned_wins} tuned')
    share = pct(base_wins / decided) if decided else '—'
    print(f'podíl pro base {share}  95% CI {interval(p_lo, p_hi)}')
    if p_lo <= 0.5 <= p_hi:
        caveat = '' if decided >= 20 else f', ale {decided} rozhodnutých dvojic ho neumí odhalit'
        print(f'→ interval obsahuje 0.5 — rozdíl neprokázán{caveat}')
    else:
        print('→ interval míjí 0.5 — rozdíl v kvalitě')

    print('\n## Rozpad podle toho, jestli kontext vůbec obsahoval odpověď')
    for with_ctx in (True, False):
        group = [(p, w) for p, w in judged if p.has_relevant_context == with_ctx]
        t = sum(1 for _, w in group if w == 'tie')
        b = sum(1 for _, w in group if w == base_model)
        u = sum(1 for _, w in group if w == tuned_model)
        name = 's relevantním kontextem ' if with_ctx else 'bez relevantního      '
        print(f'{name} n={len(group)}  base {b} : {u} tuned, remíz {t}')

    print('\n## Kontrola sleposti — šel verdikt za délkou odpovědi?')
    decided_pairs = [(p, w) for p, w in judged if w != 'tie']
    longer_won = sum(
        1
        for p, w in decided_pairs
        if len((p.base if w == base_model else p.tuned)['answer'])
        > len((p.tuned if w == base_model else p.base)['answer'])
    )
    l_lo, l_hi = wilson(longer_won, len(decided_pairs))
    print(f'delší odpověď vyhrála {longer_won}/{len(decided_pairs)}  95% CI {interval(l_lo, l_hi)}')
    verdict = (
        'nerozlišitelné od náhody — délka verdikt neřídila'
        if l_lo <= 0.5 <= l_hi
        else 'délka s verdiktem koreluje — nahlásit jako limit'
    )
    print(f'→ {verdict}')


def main() -> None:
    base_model, tuned_model = require_env('BASE_MODEL', 'TUNED_MODEL')
    all_runs: list[Run] = read_json_lines(RUNS, required=True)
    prefs: list[Preference] = read_json_lines(PREFS, required=True)
    labels: list[Labelled] = read_json_lines(DATA / 'retrieval.labels.jsonl', required=True)
    pairs = pair_up(median_across_seeds(all_runs), base_model, tuned_model, labels)
    print(f'{SET} set — n = {len(pairs)} otázek, oba modely, byte-identický vstup\n')
    report_lengths(pairs, all_runs)
    report_quality(pairs, prefs, base_model, tuned_model)


def tokens_or_nan(run: Run) -> float:
    return run['completionTokens'] if run['completionTokens'] is not None else math.nan


if __name__ == '__main__':
    main()
